rest/zone_test.go

```diff
@@ -0,0 +1,161 @@
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestPagination_zoneList(t *testing.T) {
+	// It should follow Links and gather all Zones into the list
+	testServer := mockAPI(t)
+	doer := testServer.Client()
+	client := NewClient(doer, SetEndpoint(testServer.URL))
+
+	zones, _, _ := client.Zones.List()
+
+	expected := []string{
+		"example.com",
+		"coolzone.cool",
+		"example.org",
+		"example.io",
+	}
+
+	assert.Equal(t, 4, len(zones))
+	for idx := range zones {
+		assert.Equal(t, expected[idx], zones[idx].Zone)
+	}
+}
+
+func TestPagination_zoneRecords(t *testing.T) {
+	// It should follow Links and gather all Records into the Zone
+	testServer := mockAPI(t)
+	doer := testServer.Client()
+	client := NewClient(doer, SetEndpoint(testServer.URL))
+
+	zone, _, _ := client.Zones.Get("coolzone.cool")
+
+	expected := []string{
+		"a.coolzone.cool",
+		"b.coolzone.cool",
+		"c.coolzone.cool",
+		"d.coolzone.cool",
+	}
+
+	assert.Equal(t, "coolzone.cool", zone.Zone)
+	assert.Equal(t, 4, len(zone.Records))
+	for idx := range zone.Records {
+		assert.Equal(t, expected[idx],  zone.Records[idx].Domain)
+	}
+}
+
+func TestPagination_disabled(t *testing.T) {
+	// It should not follow Links
+	testServer := mockAPI(t)
+	doer := testServer.Client()
+	client := NewClient(doer, SetEndpoint(testServer.URL), SetFollowPagination(false))
+
+	zones, _, _ := client.Zones.List()
+	expected := []string{
+		"example.com",
+		"coolzone.cool",
+	}
+	assert.Equal(t, 2, len(zones))
+	for idx := range zones {
+		assert.Equal(t, expected[idx], zones[idx].Zone)
+	}
+
+	zone, _, _ := client.Zones.Get("coolzone.cool")
+	expected = []string{
+		"a.coolzone.cool",
+		"b.coolzone.cool",
+	}
+	assert.Equal(t, "coolzone.cool", zone.Zone)
+	assert.Equal(t, 2, len(zone.Records))
+	for idx := range zone.Records {
+		assert.Equal(t, expected[idx],  zone.Records[idx].Domain)
+	}
+}
+
+const zoneListPath = "/zones"
+const zoneListQuery = "?after=coolzone.cool&limit=2"
+var zoneListLinkHeader = fmt.Sprintf(`</zones?%s>; rel="next"`, zoneListQuery)
+const zoneListPageOne = `[
+  {"zone": "example.com"},
+  {"zone": "coolzone.cool"}
+]`
+const zoneListPageTwo = `[
+  {"zone": "example.org"},
+  {"zone": "example.io"}
+]`
+
+const zoneGetPath = "/zones/coolzone.cool"
+const zoneGetQuery = "after=b.coolzone.cool&limit=2"
+var zoneGetLinkHeader = fmt.Sprintf(
+	`</zones/coolzone.cool?%s>; rel="next"`, zoneGetQuery,
+)
+const zoneGetPageOne = `{
+  "zone": "coolzone.cool",
+  "records": [{
+    "domain":"a.coolzone.cool",
+    "short_answers": ["1.1.1.1"],
+    "type":"A"
+  },
+  {
+    "domain":"b.coolzone.cool",
+    "short_answers": ["2.2.2.2"],
+    "type":"A"
+  }]
+}`
+const zoneGetPageTwo = `{
+  "zone": "coolzone.cool",
+  "records": [{
+    "domain":"c.coolzone.cool",
+    "short_answers": ["3.3.3.3"],
+    "type":"A"
+  },
+  {
+    "domain":"d.coolzone.cool",
+    "short_answers": ["4.4.4.4"],
+    "type":"A"
+  }]
+}`
```

Comment:
Are package scoped vars necessary here.  Can these live inside `mockAPI`?  Alternatively, can `mockAPI` be generalized by taking it's configuration (i.e. `http.HandlerFunc`)  from it's caller so it can be reused in the future?

Reply:
not necessary but seemed more readable than having them inside `mockApi` and not very dangerous. Definitely don't mind moving them.

I thought about generalizing `mockAPI` more, and thought about ways of organizing the data it uses better, but prefer to wait until we have some other use cases. The repo could use some more tests and I'm happy to work on some :) for now, I kinda prefer it dumb-n-hardcoded like this, it's a simple place to start when we add more tests n stuff